### PR TITLE
Move filtered registry search out of the image service

### PR DESCRIPTION
**Pull request by corhere**
- Related to #44032

Move filtered registry search logic out of the image service as querying a remote registry is unrelated to the functionality of the image service. Move the logic into the registry service as it is the source of truth for the registry configuration, which makes it the [most cohesive](https://en.wikipedia.org/wiki/Cohesion_(computer_science)) place to put it [which does not require more extensive refactoring](https://github.com/moby/moby/pull/44032#pullrequestreview-1085873256).

### CODE REVIEW

**Comment by thaJeztah** [comment]
I'm leaning towards keeping this more separate from the registry; "search" is a bit of an odd one, because search is _not part_ of the OCI distribution specification.

Search was implemented in "docker registry v1" (the Docker Index (hence `index.docker.io`), which later became Docker Hub). When the registry V2 specification was created (which became the OCI Distribution Spec), search was deliberately left out of the implementation, because indexing and searching content was a separate concern, and it became clear that "what can be searched", "how" can be searched (and what properties to search for) could be very opinionated (different providers could have different things to search for).

So search did not become part of the OCI specs.

The search functionality in the daemon is still in place _mostly_ for backward compatibility, and because the old (v1) search API is the only documented search API, so some other registry implementations also added support for it (by lack of other specifications for searching).

That said; some registries (including Docker Hub) provide other (more advanced) search functionality, and there's been some talk about defining a OCI spec for searching/indexing (not sure if that ever got actual progress though).

With the above in mind, I'd like to keep it separate with the potential to;

- Allow other implementations to be added (haven't given this much thought, but somehow making it pluggable?)
- (there's been various discussions about "better search" for images on Docker Hub at least; Docker Hub already has V2 and V3 search engines, which could provide some of that, but not sure if their API is publicly documented).  
- Probably for backward-compat, the "fallback" would be to use the V1 search

In either case; my thinking was to "try to keep it as isolated as possible"

**Comment by corhere** [comment]
> I'm leaning towards keeping this more separate from the registry; "search" is a bit of an odd one, because search is _not part_ of the OCI distribution specification.

The search interface is separated from the image distribution interface. The implementation is tied to the registry service, but that was already the case. Think of this PR as more of "getting search out of the wrong place to unblock containerd integration" than "getting search into the perfect place." We can still make #44032 happen; this PR just takes the pressure off.

**Review by rumpl — APPROVED** [review]
LGTM, a great first step, and love that it unblocks containerd a bit. Thanks

**Review by tianon — APPROVED** [review]
I agree that somewhere else is probably better, but also that "registry" seems like a more appropriate home than the image store for this Hub-specific behavior (given Hub is a registry, even if this isn't an OCI-specified endpoint). :sweat_smile:

(In other words, this seems like a very small, easy net win on the path to that larger reconsideration :heart:)

**Review by neersighted — APPROVED** [review]
Looks like a sensible cleanup/separation of concerns to me, and eases the containerd integration work. Needs a rebase, but otherwise LGTM.

**Review by vvoland — APPROVED** [review]

File: registry/search.go
Comment (thaJeztah) [inline registry/search.go:11]: This import causes the CLI to also pull in https://github.com/docker/cli/pull/4010/commits/a5d122d15c095955d04136348e4c693a76b24613 (commit vendoring is from a temporary branch, so link will likely be broken at some point);

- github.com/docker/docker/dockerversion
    - -> github.com/docker/docker/pkg/useragent
    - -> github.com/docker/docker/pkg/parsers/kernel
    - -> golang.org/x/sys/windows/registry

We'll probably need to have a look at passing the `dockerversion` / `dockerversion.DockerUserAgent(ctx)` somehow

/cc @corhere @neersighted

Comment (neersighted) [reply registry/search.go:11]: Ah, shoot, interesting.

Comment (thaJeztah) [reply registry/search.go:11]: Quick search for that;

When Authenticating;
https://github.com/moby/moby/blob/a082bbcb30c3aafd58cbec5023ac1fdb57161348/daemon/auth.go#L11-L13

Distribution sets it up as part of the transport;
https://github.com/moby/moby/blob/a082bbcb30c3aafd58cbec5023ac1fdb57161348/distribution/registry.go#L100-L101

And for plugins;
https://github.com/moby/moby/blob/a082bbcb30c3aafd58cbec5023ac1fdb57161348/plugin/registry.go#L30-L34

So looks like it's a bit "all over the place"

**Pull request merged**
